### PR TITLE
Modify the main page of the application #28 

### PR DESCRIPTION
**Pull request by Farcasut**
Modified feature:
- Modified the main page of the application to be a dashboard.
-  Removed the old analytics dashboard class/template/provider

New Features:
- Added a filtering macro

Preview:
image

### CODE REVIEW

**Comment by Farcasut** [comment]
image

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filtering.xml [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+  #set ($placeholderStart = $request.analyticsStartDate)
+#end
+#if($request.analyticsEndDate &amp;&amp; $request.analyticsEndDate != '')
+  #set ($placeholderEnd = $request.analyticsEndDate)
+#end
+{{html clean="false"}}
+&lt;form id="analyticsFilter" class="form-horizontal"&gt;
+  &lt;div class="form-group"&gt;
+    &lt;!-- Start Date Label --&gt;
+    &lt;label for="analyticsStartDate" class="col-sm-2 control-label"&gt;$services.localization.render('rendering.macro.filtering.startDate')&lt;/label&gt;
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filtering.xml:301]: You should add a title for this and the end date label. Maybe get inspired by Matomo 

Also, line exceeds the 120ch limit, below as well

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml [hunk 2]
```diff
@@ -38,25 +38,24 @@
   <hidden>true</hidden>
   <content>{{velocity output='false'}}
 #macro (getDate)
-  #set ($dashbord = $doc.getObject('Analytics.Code.AnalyticsDashboardClass'))
   ## A dashboard will have a date inside and that date will take precedence over the date set by the macro.
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml:41]: This could be updated as well

File: application-analytics-ui/src/main/resources/Analytics/WebHome.xml [hunk 3]
```diff
@@ -27,45 +27,187 @@
...
-  <parent>Main.WebHome</parent>
+  <parent>xwiki:Main.WebHome</parent>
   <author>xwiki:XWiki.Admin</author>
   <contentAuthor>xwiki:XWiki.Admin</contentAuthor>
   <version>1.1</version>
   <title>Analytics</title>
   <comment/>
   <minorEdit>false</minorEdit>
-  <syntaxId>xwiki/2.1</syntaxId>
+  <syntaxId>xwiki/2.0</syntaxId>
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/WebHome.xml:37]: Why is this needed?

File: application-analytics-ui/src/main/resources/Analytics/WebHome.xml [hunk 4]
```diff
@@ -27,45 +27,187 @@
...
-{{/velocity}}</content>
+  <content>{{dashboard/}}
+  </content>
+  <object>
+    <name>Analytics.WebHome</name>
+    <number>0</number>
+    <className>XWiki.DocumentSheetBinding</className>
+    <guid>c9b959c5-3c33-4cf0-8598-1fb3e023213c</guid>
+    <class>
+      <name>XWiki.DocumentSheetBinding</name>
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/WebHome.xml:47]: Is this needed? I see it is empty

File: application-analytics-ui/src/main/resources/Analytics/WebHome.xml [hunk 5]
```diff
@@ -27,45 +27,187 @@
   <defaultLanguage>en</defaultLanguage>
   <translation>0</translation>
   <creator>xwiki:XWiki.Admin</creator>
-  <parent>Main.WebHome</parent>
+  <parent>xwiki:Main.WebHome</parent>
   <author>xwiki:XWiki.Admin</author>
   <contentAuthor>xwiki:XWiki.Admin</contentAuthor>
   <version>1.1</version>
   <title>Analytics</title>
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/WebHome.xml:34]: This page should be excluded from the licensing check, in order for users to be able to edit it.
Because of this, the display is also broken. On your instance you have the page created after application install, so the license check is not applied
image

**Review by oanalavinia — CHANGES_REQUESTED** [review]

File: application-analytics-ui/src/main/resources/Analytics/WebHome.xml (hunk 3 above)
Comment (Farcasut) [reply application-analytics-ui/src/main/resources/Analytics/WebHome.xml:37]: When I created the new homepage using the dashboard template, it also changed this property.

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/MostViewedPages.xml [hunk 6]
```diff
@@ -241,8 +241,12 @@
     </property>
     <property>
       <code>{{include reference='Analytics.Code.Macros.VelocityMacros'/}}
+{{include reference="Licenses.Code.VelocityMacros"/}}
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/MostViewedPages.xml:244]: New line missing

File: application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsTranslations.xml [hunk 7]
```diff
@@ -99,7 +100,11 @@ analytics.livedata.action.rowEvolution.title=Row Evolution
 ## Others
 analytics.extension.name=Analytics Application (Pro)
 analytics.extension.title=Analytics
-</content>
+rendering.macro.analyticsFilter.startDate=Start date
+rendering.macro.analyticsFilter.endDate=End date
+rendering.macro.analyticsFilter.apply=Apply
+rendering.macro.analyticsFilter.name=Analytics Filter
+rendering.macro.analyticsFilter.description=Updates all the analytics macros on the same page as this macro to the period set by the user.</content>
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsTranslations.xml:107]: This needs some reformulations, it's not very clear
`Add filtering options for the analytics macros of the current page. The filtering criteria will be saved in the current URL, to have a reproducible state, and will also override eventual filter set at macro level.` - probably still needs improvements

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml [hunk 8]
```diff
@@ -0,0 +1,501 @@
...
+    </property>
+    <property>
+      <async_enabled>0</async_enabled>
+    </property>
+    <property>
+      <code>{{velocity}}
+#set ($discard =  $xwiki.ssx.use('Analytics.Code.Macros.Filter'))
+#if ($services.licensing.licensor.hasLicensureForEntity(
+    $services.model.createDocumentReference('', ['Analytics', 'Code'], 'ConfigurationClass')))
+#set ($placeholderStart = '2007-01-01')
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml:414]: Identation. 
Here as well, you can also move all code inside a macro, so to not mix licensing check with other macro specific things. E.g. https://github.com/xwikisas/xwiki-pro-macros/blob/main/xwiki-pro-macros-ui/src/main/resources/Confluence/Macros/Tip.xml#L283

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml [hunk 9]
```diff
@@ -0,0 +1,501 @@
...
+#set ($placeholderEnd = $datetool.get('yyyy-MM-dd'))
+#if($request.analyticsStartDate &amp;&amp; $request.analyticsStartDate != '')
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml:416]: Space missing after `if`, below as well

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml [hunk 10]
```diff
@@ -0,0 +1,501 @@
...
+  &lt;div class="row"&gt;
+    &lt;div class="col-sm-2"&gt;
+      &lt;!-- Start Date Label &amp; Input Group --&gt;
+      &lt;div&gt;
+        &lt;label for="analyticsStartDate"&gt;$services.localization.render('rendering.macro.analyticsFilter.startDate')&lt;/label&gt;
+        #set ($dateTimePickerParams = {
+        'id': 'analyticsStartDate',
+        'name': 'analyticsStartDate',
+        'data-format': 'yyyy-MM-dd',
+        'placeholder': $placeholderStart
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml:433]: Identation. Below as well

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml [hunk 11]
```diff
@@ -0,0 +1,501 @@
...
+        })
+        #dateTimePicker($dateTimePickerParams)
+      &lt;/div&gt;
+    &lt;/div&gt;
+    &lt;!-- Submit Button --&gt;
+    &lt;div class="col-sm-2"&gt;
+      &lt;div&gt;
+        &lt;div&gt;
+          &amp;nbsp;
+        &lt;/div&gt;
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml:456]: This is not ok

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml [hunk 12]
```diff
@@ -38,25 +38,24 @@
   <hidden>true</hidden>
   <content>{{velocity output='false'}}
 #macro (getDate)
-  #set ($dashbord = $doc.getObject('Analytics.Code.AnalyticsDashboardClass'))
-  ## A dashboard will have a date inside and that date will take precedence over the date set by the macro.
-  #if ($dashbord)
-    #set ($startDate = $dashbord.getProperty('startDate'))
-    #set ($endDate = $dashbord.getProperty('endDate'))
+  ## A dashboard will have the ability to add parameters to the url to filter all the macros in the page.
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml:41]: `If the analytics filter is used on a dashboard page, this could add filter parameters in the URL and these have priority over what is set at the macro level.`

File: application-analytics-ui/src/main/resources/Analytics/WebHome.xml (hunk 3 above)
Comment (oanalavinia) [reply application-analytics-ui/src/main/resources/Analytics/WebHome.xml:37]: Can you check if there are problems with using `2.1`?

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/MostViewedPages.xml [hunk 13]
```diff
@@ -241,8 +241,12 @@
...
 
 {{velocity}}
+#if ($services.licensing.licensor.hasLicensureForEntity(
+    $services.model.createDocumentReference('', ['Analytics', 'Code'], 'ConfigurationClass')))
+
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/MostViewedPages.xml:249]: There is an extra newline displaye dabove the mostViewedPages macro, and it might come from here

**Review by oanalavinia — CHANGES_REQUESTED** [review]
The filter UI might need some improvements for orientation, hints. But we can leave it like this for the moment and update it after getting some feedback

Not sure if it's also related to this changes, but the row evolution button is not working on this branch (tried in incognito)

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml (hunk 11 above)
Comment (Farcasut) [reply application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml:454]: Oki! I will try to find another way to align the apply button

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/RowEvolution.xml [hunk 14]
```diff
@@ -254,6 +254,11 @@ require(['jquery', 'chart'], ($) =&gt; {
     // This rowIdentifier is used by Matomo to identify the row for which you want to retrieve the data.
     const rowIdentifier = $(this).data('rowIdentifier');
     currentModal = $(this).closest('.xwiki-livedata').next(SELECTORS.modal);
+    // On XWiki 14+ the structure of the dom was changed and I have to check if an element was found.
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/RowEvolution.xml:257]: `The livedata structure has changed starting with XWiki 14+ and so the modal will be undefined on these version. Remove this check after upgrading to the 14.10 XWiki parent.`

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/RowEvolution.xml [hunk 15]
```diff
@@ -254,6 +254,11 @@ require(['jquery', 'chart'], ($) =&gt; {
...
+    if(!currentModal.length) {
+    // Selector for XWiki 14+
+    currentModal=$(this).closest('.liveData').next(SELECTORS.modal);
+    }
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/RowEvolution.xml:261]: Codestyle

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/MostViewedPages.xml [hunk 16]
```diff
@@ -242,50 +242,60 @@
...
-      {'id': 'bounceRate', 'displayer': 'html'},
-      {'id': 'exitRate', 'displayer': 'html'},
-      {'id': 'actions', 'filterable': false, 'sortable': false, 'displayer': 'html'}
-    ],
-    'entryDescriptor': {
-      'idProperty': 'title'
+#macro(executeMacro)
+  #getDate()
+  #set ($discard =$xwiki.jsx.use('Analytics.Code.Macros.RowEvolution'))
+  #set ($period = 'range')
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/MostViewedPages.xml:251]: Why do you need a separate variable?

File: application-analytics-ui/src/main/resources/Analytics/Code/Configuration.xml [hunk 17]
```diff
@@ -385,4 +385,4 @@
       <sectionOrder/>
     </property>
   </object>
-</xwikidoc>
+</xwikidoc>
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Configuration.xml:388]: You forgot to apply xar:format

**Review by oanalavinia — CHANGES_REQUESTED** [review]

File: application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsTranslations.xml [hunk 18]
```diff
@@ -55,11 +55,12 @@ analytics.mostViewedPages.header.emptyvalue=-
 analytics.mostViewedPages.header.exitRate=Exit rate
 analytics.mostViewedPages.header.hits=Page views
 analytics.mostViewedPages.header.timeSpent=Time spent(s)
-analytics.mostViewedPages.header.title=Page title
+analytics.mostViewedPages.header.pageTitle=Page title
 analytics.mostViewedPages.header.visits=Unique visitors
 analytics.mostViewedPages.header.year=Year
 analytics.mostViewedPages.parameter.label.startDate=Start time of the date interval used for this macro
 analytics.mostViewedPages.parameter.label.endDate=End time of the date interval used for this macro
+rendering.macro.mostViewedPage.name=Most viewed pages
```
[the remark this answers is not in the document]
Comment (Farcasut) [reply application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsTranslations.xml:63]: I will change the name of the macro from mostViwedPage -> mostViwedPages into a separate pr as you suggested in the initial comment.

[the remark this answers is not in the document]
Comment (oanalavinia) [reply application-analytics-ui/src/main/resources/Analytics/Code/AnalyticsTranslations.xml:63]: Ok. Please open an issue for this

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml [hunk 19]
```diff
@@ -0,0 +1,507 @@
...
+  #set ($placeholderStart = $request.analyticsStartDate)
+#end
+#if ($request.analyticsEndDate &amp;&amp; $request.analyticsEndDate != '')
+  #set ($placeholderEnd = $request.analyticsEndDate)
+#end
+{{html clean="false"}}
+&lt;div class="xform half"&gt;
+	&lt;form id="analyticsFilter" class="form-flex"&gt;
+		&lt;div class="form-item flex-item"&gt;
+			&lt;!-- Start Date Label &amp; Input Group --&gt;
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml:435]: Indentation is with 4 spaces instead of 2. Check at the end, as I think there are other indentation rules broken by (most probably) your IDEA

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml [hunk 20]
```diff
@@ -0,0 +1,507 @@
...
+			&lt;label for="analyticsStartDate"&gt;$services.localization.render('rendering.macro.analyticsFilter.startDate')&lt;/label&gt;
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml:436]: Because translations can be added by users, besides the ones that come protected by a license, you need to make sure that they don't introduce security issues. So always wrap the `$services.localization.render` in a `$escapetool.xml` (check the pages for where you need this)

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml [hunk 21]
```diff
@@ -0,0 +1,507 @@
...
+        <unmodifiable>0</unmodifiable>
+        <values>currentPage|onDemand|always</values>
+        <classType>com.xpn.xwiki.objects.classes.StaticListClass</classType>
+      </use>
+    </class>
+    <property>
+      <cache>long</cache>
+    </property>
+    <property>
+      <code>.form-flex {
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml:187]: These class names are too generic and might crash with other classes

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml [hunk 22]
```diff
@@ -0,0 +1,507 @@
...
+  display: flex;
+  flex-wrap: wrap;
+}
+.form-item {
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml:191]: Why you need 2 separated classes for an item?

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/RowEvolution.xml [hunk 23]
```diff
@@ -254,6 +254,12 @@ require(['jquery', 'chart'], ($) =&gt; {
     // This rowIdentifier is used by Matomo to identify the row for which you want to retrieve the data.
     const rowIdentifier = $(this).data('rowIdentifier');
     currentModal = $(this).closest('.xwiki-livedata').next(SELECTORS.modal);
+    // The livedata structure has changed starting with XWiki 14+ and so the modal will be undefined on these
+    // version. Remove this check after upgrading to the 14.10 XWiki parent.
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/RowEvolution.xml:258]: versions*

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/RowEvolution.xml [hunk 24]
```diff
@@ -254,6 +254,12 @@ require(['jquery', 'chart'], ($) =&gt; {
...
+    if(!currentModal.length) {
+      // Selector for XWiki 14+
+      currentModal=$(this).closest('.liveData').next(SELECTORS.modal);
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/RowEvolution.xml:261]: Just some spaces needed + a `.` for your comment

**Review by oanalavinia — CHANGES_REQUESTED** [review]
Almost:)

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml (hunk 22 above)
Comment (Farcasut) [reply application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml:191]: Not really I thought that it would be much cleaner to separate the css into 2 classes.

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml (hunk 20 above)
Comment (Farcasut) [reply application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml]: Should I replace all the instances of $services,localization.render with a regex just to be safe?

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml [hunk 25]
```diff
@@ -0,0 +1,507 @@
...
+#if ($request.analyticsEndDate &amp;&amp; $request.analyticsEndDate != '')
+  #set ($placeholderEnd = $request.analyticsEndDate)
+#end
+{{html clean="false"}}
+&lt;div class="xform half"&gt;
+  &lt;form id="analyticsFilter" class="analytics-form-flex"&gt;
+	&lt;div class="analytics-form-item analytics-flex-item"&gt;
+	  &lt;!-- Start Date Label &amp; Input Group --&gt;
+      &lt;label for="analyticsStartDate"&gt;$escapetool.xml($services.localization.render('rendering.macro.analyticsFilter.startDate'))&lt;/label&gt;
+		#set ($dateTimePickerParams = {
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml:437]: This is not inside the label tag, so there is no need for it being with 2 spaces on the right, but it should be at the same indentation

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml [hunk 26]
```diff
@@ -0,0 +1,507 @@
...
+  #set ($placeholderStart = $request.analyticsStartDate)
+#end
+#if ($request.analyticsEndDate &amp;&amp; $request.analyticsEndDate != '')
+  #set ($placeholderEnd = $request.analyticsEndDate)
+#end
+{{html clean="false"}}
+&lt;div class="xform half"&gt;
+  &lt;form id="analyticsFilter" class="analytics-form-flex"&gt;
+	&lt;div class="analytics-form-item analytics-flex-item"&gt;
+	  &lt;!-- Start Date Label &amp; Input Group --&gt;
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/Filter.xml:435]: Maybe `Start date label and input`?

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml [hunk 27]
```diff
@@ -38,25 +38,24 @@
   <hidden>true</hidden>
   <content>{{velocity output='false'}}
 #macro (getDate)
-  #set ($dashbord = $doc.getObject('Analytics.Code.AnalyticsDashboardClass'))
-  ## A dashboard will have a date inside and that date will take precedence over the date set by the macro.
-  #if ($dashbord)
-    #set ($startDate = $dashbord.getProperty('startDate'))
-    #set ($endDate = $dashbord.getProperty('endDate'))
+  ## If the analytics filter is used on a dashboard page, this could add filter parameters in the URL and these have priority over what is set at the macro level.
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml:41]: Line too long

File: application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml [hunk 28]
```diff
@@ -156,7 +155,7 @@
     &lt;div class="container-fluid"&gt;
       &lt;div class="row"&gt;
         &lt;div class="col-md-3"&gt;
-          &lt;label for="rowEvolutionPeriod$id"&gt;$services.localization.render('analytics.rowEvolution.modal.period')
+          &lt;label for="rowEvolutionPeriod$id"&gt;$escapetool.xml($services.localization.render('analytics.rowEvolution.modal.period'))
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/Code/Macros/VelocityMacros.xml:158]: Exceeds 120ch now. Below as well

File: application-analytics-ui/src/main/resources/Analytics/WebHome.xml [hunk 29]
```diff
@@ -199,9 +340,9 @@
       <name>platform.panels.AnalyticsApplication</name>
     </property>
     <property>
-      <parameters>label=$services.localization.render('analytics.extension.title')
-target=Analytics.WebHome
-icon=icon:chart-bar</parameters>
+      <parameters>label=$escapetool.xml($services.localization.render('analytics.extension.title'))
```
Comment (oanalavinia) [inline application-analytics-ui/src/main/resources/Analytics/WebHome.xml:343]: No need to do this here as well, since it's escaped later on https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-application/xwiki-platform-application-ui/src/main/resources/Applications/Macros.xml#L50

**Review by oanalavinia — CHANGES_REQUESTED** [review]

**Review by oanalavinia — APPROVED** [review]
